### PR TITLE
Knitr changes

### PR DESCRIPTION
**Pull request by methodds**
I adjusted the backend in the S3 printfunction, so that it now uses ```htmltools::HTML```. According to the rmarkdown guys [see here](https://github.com/rstudio/rmarkdown/issues/964#issuecomment-279398894) this should also work for r notebooks.

### CODE REVIEW

**Comment by methodds** [comment]
Just to let you know:  the htmltools::HTML solution is still something I'm working on, as it doesn't behave as intended at the moment.

File: R/sjTabItemAnalysis.R [hunk 1]
```diff
@@ -92,24 +92,22 @@
...
-#'  
+#' 
 #' # create data frame with COPE-index scale
 #' mydf <- data.frame(efc[, c(start:end)])
 #' colnames(mydf) <- varlabs[c(start:end)]
-#' 
-#' \dontrun{
-#' sjt.itemanalysis(mydf)
+#' #' #'
+#' #' #' \dontrun{
```
Comment (sjPlot) [inline R/sjTabItemAnalysis.R]: Why are you uncommenting this?

File: R/sjTabLinReg.R [hunk 2]
```diff
@@ -143,21 +143,9 @@
...
-#' fit2 <- lm(neg_c_7 ~ c160age + c12hour + c161sex + c172code, data = efc)
-#' 
-#' # create and open HTML-table in RStudio Viewer Pane or web browser
-#' # note that we don't need to specify labels for the predictors,
-#' # because these are automatically read
-#' sjt.lm(fit1, fit2)
-#' 
-#' # create and open HTML-table in RStudio Viewer Pane or web browser
+#' models below at the end.
+#' library(sjmisc)# data(efc)# # # fit first model# fit1 <- lm(barthtot ~ c160age + c12hour + c161sex + c172code, data = efc)# # fit second model# fit2 <- lm(neg_c_7 ~ c160age + c12hour + c161sex + c172code, data = efc)# # # create and open HTML-table in RStudio Viewer Pane or web browser# # note that we don't need to specify labels for the predictors,# # because these are automatically read# sjt.lm(fit1, fit2)
```
Comment (sjPlot) [inline R/sjTabLinReg.R]: linebreaks lost, probably revert these changes?

File: R/sjTabItemAnalysis.R (hunk 1 above)
Comment (methodds) [inline R/sjTabItemAnalysis.R]: Sorry this is a mistake. My workflow was basically uncomment examples, use them for testing, comment out again. I guess in this case I forgot that. Is there a better way for testing?

File: R/sjTabItemAnalysis.R [hunk 3]
```diff
@@ -372,14 +370,23 @@ sjt.itemanalysis <- function(df,
...
+      knitr = knitr,
+      file = file,
+      show = !no.output,
+      use.viewer = use.viewer,
+      df.list = df.ia,
+      index.scores = index.scores,
+      df.index.scores = df.index.scores,
+      cronbach.values = cronbach.total,
+      ideal.item.diff = diff.ideal.list,
+      knitr.list = knitr.list
```
Comment (sjPlot) [inline R/sjTabItemAnalysis.R]: I have to check what knitr.list actually is needed for, so it might be integrated in print-method.

File: R/sjTabMannWhitney.R [hunk 4]
```diff
@@ -26,10 +26,10 @@
 #'
 #' @examples
 #' \dontrun{
-#' library(sjmisc)
-#' data(efc)
+#' #' library(sjmisc)
+#' #' data(efc)
```
Comment (sjPlot) [inline R/sjTabMannWhitney.R]: Why uncommenting this?

File: vignettes/sjtbasic.Rmd [hunk 5]
```diff
@@ -77,31 +77,16 @@ Most `sjt` function invisibly return a structure with following values:
 * the complete html-output `output.complete`
 * the html-table with inline-css for use with knitr `knitr`
 
-`page.style` and `page.content` are described below in the CSS section. If you want to include a HTML table in your knitr document, simply use an inline code chunk. Be sure to use `no.output=TRUE`, because else your output would also be opened in a viewer pane or webbrowser.
+`page.style` and `page.content` are described below in the CSS section. If you want to include a HTML table in your knitr document, simply use the corresponding function within a code chunk. Be sure to use `no.output=TRUE`, because else your output would also be opened in a viewer pane or webbrowser.
 
 `r rinline("sjt.frq(efc\\$e42dep, no.output=TRUE)\\$knitr")`
```
Comment (sjPlot) [inline vignettes/sjtbasic.Rmd]: I think we don't need line 82 any more.

File: vignettes/sjtbasic.Rmd [hunk 6]
```diff
@@ -151,10 +139,10 @@ sjt.frq(efc$e42dep,
                    css.lasttablerow = 'border-top: 1px solid; border-bottom: none;', 
                    css.summary = 'color: blue;'))
 ```
-`r sjt.frq(efc$e42dep, no.output=TRUE, CSS=list(css.centeralign='text-align: left;', css.caption='font-weight: normal; font-style: italic;', css.firsttablecol='font-weight: bold;', css.lasttablerow='border-top: 1px solid; border-bottom: none;', css.summary='color: blue;'))$knitr`
```
Comment (sjPlot) [inline vignettes/sjtbasic.Rmd]: By removing this, `eval = FALSE` in line 134 must be removed. Same for all eval=FALSE in following code-chunks, since these code-chunks can now be evaluated.

**Comment by sjPlot** [comment]
I can't commit and sync from here (my Windows machine), so I have to wait until I'm back home - I think, the changes should work. Many thank for this! You should add yourself as contributor to the DESCRIPTION-file. :-)

File: R/sjTabItemAnalysis.R (hunk 1 above)
Comment (sjPlot) [inline R/sjTabItemAnalysis.R]: You could write tests with package *testthat* ;-) But, copy/paste from the docs to console works, you only need to uncomment when doing R CMD check. However, opening web browsers in examples is not good practice according CRAN guidelines, hence I added the dont-run-tag to all HTML-functions.

File: R/sjTabMannWhitney.R (hunk 4 above)
Comment (methodds) [inline R/sjTabMannWhitney.R]: See above,  I guess my (sloppy executed)  way of testing is probably not the best one =D

**Comment by methodds** [comment]
Thanks for the offer, will gladly do that tomorrow.  I'm however still waiting for feedback from the R markdown guys, as I would really want to the changes to also work within R-notebooks.

File: vignettes/sjtbasic.Rmd (hunk 5 above)
Comment (sjPlot) [inline vignettes/sjtbasic.Rmd]: Maybe the line around `rinline <- function(code){...` neither. I'll check that at home.

**Comment by sjPlot** [comment]
Are you addressing my comments, or do you want me to merge your PR and fix them?

**Comment by methodds** [comment]
I can do this until tomorrow the latest no problem, but I think that as long as the [r-notebook problems](https://github.com/rstudio/rmarkdown/issues/964#issuecomment-279650639) aren't fixed, we should stick to the ```knitr::asis_output()``` method for printing.

**Comment by sjPlot** [comment]
No, there's no time pressure, I just wanted to know who's next to work in this PR ;-)
 Does `htmltools::HTML()` have any disadvantages? If it works fine within knitr, I don't see any need to change it.

**Comment by methodds** [comment]
Did you try it with r-notebooks? Because in the few tests that I ran on them it did not work properly. The corresponding notebook previews didn't even show the commands, e.g. ```sjt.frq(foo)```

**Comment by sjPlot** [comment]
No, not tried yet. I would merge this PR and sync with my desktop to check the changes, so I'll wait until you "release" this PR for merging - notebook-support can be added in another PR. If knitr-integration is already fixed by this PR, it's fine.

File: R/sjTabItemAnalysis.R (hunk 1 above)
Comment (sjPlot) [inline R/sjTabItemAnalysis.R]: I think my answer was a bit confusing. The dontrun-tag has no consequences, when copy&paste examples from docs to console. It's just that example sourrounded by dontrun are not run inside CRAN checks (R CMD check). You don't need to comment/uncomment to test these examples (unless you want to test using R CMD check).

File: R/sjTabItemAnalysis.R (hunk 3 above)
Comment (sjPlot) [inline R/sjTabItemAnalysis.R]: Ok, knitr-list are just the single elements of the complete output - so, it's not needed in the S3-print-method.

File: vignettes/sjtbasic.Rmd (hunk 5 above)
Comment (sjPlot) [inline vignettes/sjtbasic.Rmd]: I think this vignette needs some more revisions, I will do this after merging the PR.

**Comment by methodds** [comment]
I actually never released a PR, how does this work? Can I just hit "close pull request" and all the changes you suggested are applied automatically?

**Comment by sjPlot** [comment]
Yes, after you finished with all your changes you would like to apply, just tell me when you're done and I'll merge it.

**Comment by methodds** [comment]
Ok I see thanks. I got just one more question: What's the best way to add contributors to a package? I saw that there is a [ predefined format](http://r-pkgs.had.co.nz/description.html#author) for this, but the line in your Description file does not match that format (role is missing):
```Author: Daniel Lüdecke <d.luedecke@uke.de>``` Can I just manually adjust this to 
```Author: Daniel Lüdecke <d.luedecke@uke.de>, Carsten Schwemmer <carsten.schwemmer@uni-bamberg.de> [ctb]```
?
or maybe ```Author: Daniel Lüdecke <d.luedecke@uke.de> [aut, cre], Carsten Schwemmer <carsten.schwemmer@uni-bamberg.de> [ctb]```

**Comment by sjPlot** [comment]
I would say, you could follow this way:
https://github.com/hadley/dplyr/blob/master/DESCRIPTION
https://github.com/lme4/lme4/blob/master/DESCRIPTION

And your role is probably "contributor" (ctb)?

**Comment by methodds** [comment]
Okay so after learning that it is a wise idea to pull before you push, I hope that the last commit addressed all of your comments (except the vignette changes that you will probably implement on your own). I also reversed the s3 function back to ```knitr::asis_output()``` until I come up with a solution for rnotebooks.
Oh and I'm not sure how exactly your versioning works, so I just changed the date and incremented to 2.3.0.9001. If you don't find any other issues, feel free to merge this.

**Comment by sjPlot** [comment]
Well, according to versioning, I try to follow [semVer](http://semver.org/), however, sometimes it is difficult to decide what is more a patch, and what more a minor update. And the `.9000` is added by `devtools::use_dev_version()` - this is mainly, because if you install from GitHub version 2.3.1 and then it's officially updated on CRAN, CRAN won't recognize this as update. The GitHub version suffix `.9000` does not need to be changed when committing new things to GitHub, but it doesn't matter, if you increase it, either.

So, I'll merge branches now, right?

**Comment by methodds** [comment]
Yep, feel free to merge it. And thanks for the patience. the next time I'll have more routine with this :)

**Comment by sjPlot** [comment]
btw, there are (at least) three common things in modern programming languages, which I'll probably never fully understand:

1. regular expressions
2. make-file / ANT-build-scripts
3. How to deal with different branches, e.g. here in Git

That's why I usually always work with the master branch. :-)

**Comment by sjPlot** [comment]
Do you use RStudio?

**Comment by methodds** [comment]
yep and I used an [online guide](http://r-bio.github.io/intro-git-rstudio/) to set up the PR. As far as I can tell RStudio is the best IDE for R. 
In earlier days I mostly used [jupyter notebooks](https://github.com/IRkernel/IRkernel), as I'm used to working with them in Python, but the development for R is kinda stuck there.

**Comment by sjPlot** [comment]
ok, I just wondered what settings you have, because of all the committed files with no changes. I always have this problem when switching between mac and windows.

What are your settings under "Tools" -> "Global Options" -> "Code" -> "Saving"?

Mine are:
- file ends with new line - checked
- strip whitespace - checked
- LF platform native
- encoding UTF-8

**Comment by methodds** [comment]
- file ends with new line - unchecked
- strip whitespace - unchecked
- LF platform native
- encoding UTF-8

Did you try unchecking "file ends with newline"? Because I came across an annoying issue related to webcrawling. Basically, windows uses another code for newlines than linux and I also think mac, which you can see here:
http://www.cs.toronto.edu/~krueger/csc209h/tut/line-endings.html
Maybe this is the problem?

**Comment by sjPlot** [comment]
Another issue, when running R CMD check:

```
* checking for missing documentation entries ... WARNING
Undocumented code objects:
  ‘knit_print.sjTable’
All user-level objects in a package should have documentation entries.
```

I'm not sure why, because we have implemented this method exactly as suggested in the knitr-docs.

**Comment by sjPlot** [comment]
Maybe R has changed its "policies" in this issue. I followed the 2nd suggestion and imported knitr's `knit_print`, so now everything works fine. Vignettes are also revised now.

So, this PR can be closed. Feel free to look at the list of issues and work on the next PR :-D

**Comment by methodds** [comment]
Okay :) Then I'll delete the ``knitr-changes`` branch

**Comment by sjPlot** [comment]
To not forget this reference, I link to this issue:
https://github.com/gforge/htmlTable/issues/26

Since Max seems to face similar problems, we should track of his possible solutions.

**Comment by sjPlot** [comment]
Looks like R-Notebooks are still too buggy for serious use. If I change the code, the output is still the old one in the preview etc. In my impression, R notebooks currently only work for very specific situations, but are not usable in all-day-work. Maybe we should not yet take too much effort on R notebooks compatibility.

**Comment by methodds** [comment]
Which is pretty sad, because the idea is awesome :(. I'll keep an eye on the notebook development though.
